### PR TITLE
Added proxying support for iOS

### PR DESCRIPTION
**Pull request by gitlaura**
- Updated plugin.xml, README, package.json and to include ios platform
- Added ChromeProxy.m and ChromeProxy.h to respond to cordova commands and set up a proxy server
- Added ChromeProxyURLProtocol.m and ChromeProxyURLProtocol.h as a subclass of NSURLProtocol to intercept URL Requests when a proxy server is setup

### CODE REVIEW

File: src/ios/ChromeProxy.m [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+static NSDictionary *proxyDictionary;
+
++ (void)load {
+    NSLog(@"%@ Registering nsurlprotocol class", LOG_TAG);
+    [NSURLProtocol registerClass:[ChromeProxyURLProtocol class]];
+}
+
++ (void)setProxy:(NSString *)scheme withHost:(NSString *)host withPort:(NSNumber *)port {
+    proxyDictionary = @{
+                        (NSString *)[NSString stringWithFormat:@"%@Enable", scheme]:[NSNumber numberWithInt:1],
```
Comment (avidrissman) [inline src/ios/ChromeProxy.m]: Why not @YES? At least @1.

File: src/ios/ChromeProxy.m [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+                        (NSString *)[NSString stringWithFormat:@"%@Proxy", scheme]:host,
+                        (NSString *)[NSString stringWithFormat:@"%@Port", scheme]:port,
```
Comment (avidrissman) [inline src/ios/ChromeProxy.m]: +[NSString stringWithFormat:] returns an NSString*; I don't understand your cast here and in the previous two lines.

File: src/ios/ChromeProxy.m [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+}
+
+- (void)sendPluginResult:(CDVInvokedUrlCommand *)command withError:(NSString *)errorMessage {
+    CDVPluginResult* pluginResult;
+    if (errorMessage) {
+        pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:errorMessage];
+    } else {
+        pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
+    }
+    [pluginResult setKeepCallback:[NSNumber numberWithBool:YES]];
```
Comment (avidrissman) [inline src/ios/ChromeProxy.m]: [pluginResult setKeepCallback:@YES];

?

**Comment by avidrissman** [comment]
BTW, @bemasc asked me to comment. I don't know Cordova so I can't comment on whether this is right, so this is more of a general ObjC review.

**Comment by gitlaura** [comment]
Thanks @avidrissman! I made those changes.

**Comment by avidrissman** [comment]
👍

LGTM

File: src/ios/ChromeProxy.h [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+ NSURLSessionConfiguration's connectionProxyDictionary
+ in ChromeProxyURLProtocol.
+ */
++ (NSDictionary *)proxyDictionary;
+
+/* Sets the scheme, host, and port for proxyDictionary. */
++ (void)setProxy:(NSString *)scheme withHost:(NSString *)host withPort:(NSNumber *)port;
+
+@end
+
```
Comment (bemasc) [inline src/ios/ChromeProxy.h]: Excess blank lines

File: src/ios/ChromeProxy.m [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+}
+
+- (void)set:(CDVInvokedUrlCommand *)command {
+    NSLog(@"%@ Setting proxy", LOG_TAG);
+    
+    NSString* errorMessage;
+    NSDictionary* commandValue = [command argumentAtIndex:0];
+    NSDictionary* singleProxy = commandValue[@"value"][@"rules"][@"singleProxy"];
+    
+    if(!singleProxy){
```
Comment (bemasc) [inline src/ios/ChromeProxy.m]: Might as well have `if (` for consistency

File: src/ios/ChromeProxy.m [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+        } else {
+            errorMessage = @"Proxy scheme must be socks5 or http";
+            return [self sendPluginResult:command withError:errorMessage];
+        }
+    } else {
+        errorMessage = @"Failed to parse singleProxy for host and port";
+        return [self sendPluginResult:command withError:errorMessage];
+    }
+    
+    [self sendPluginResult:command];
```
Comment (bemasc) [inline src/ios/ChromeProxy.m:70]: We should at least have a TODO for `bypassList` support.

File: README.md [hunk 7]
```diff
@@ -13,6 +13,8 @@ API.  Only the "fixed_servers" and "direct" modes are supported.  Only the
 "http" and "socks5" schemes are supported, and the bypassList setting is only
 supported for the "http" scheme.
 
+On iOS, only the "http" and "socks5" schemes are supported.
```
Comment (bemasc) [inline README.md:16]: Probably should mention that `bypassList` is not supported.

File: src/ios/ChromeProxy.h (hunk 4 above)
Comment (gitlaura) [inline src/ios/ChromeProxy.h]: Fixed!

File: src/ios/ChromeProxy.m (hunk 5 above)
Comment (gitlaura) [inline src/ios/ChromeProxy.m]: Good catch. Done.

File: src/ios/ChromeProxy.m (hunk 6 above)
Comment (gitlaura) [inline src/ios/ChromeProxy.m:70]: Done

File: README.md (hunk 7 above)
Comment (gitlaura) [inline README.md:16]: Done

**Comment by bemasc** [comment]
:+1:

**Pull request merged**